### PR TITLE
Added show default constructor feature

### PR DESCRIPTION
**Pull request by cathylee12**
Follow up questions:
I haven't tested the UI configuration where the user can check the elements display options.
I added the option to ParserConfigure.java, but I did not go beyond that yet.
Also, I think it makes more sense if the showDefaultConstructor should only work if showConstructors is enabled.
I need your opinion on this.
Thank you!

### CODE REVIEW

**Comment by espertus** [comment]
> I haven't tested the UI configuration where the user can check the elements display options.

Yes. I don't know how to build the plugin to test that. I might leave that to the canonical repo owner.

> Also, I think it makes more sense if the showDefaultConstructor should only work if showConstructors is enabled.
I need your opinion on this.

I agree.

File: plantuml-parser-core/src/main/java/com/shuzijun/plantumlparser/core/ClassVoidVisitor.java
Comment (espertus) [inline plantuml-parser-core/src/main/java/com/shuzijun/plantumlparser/core/ClassVoidVisitor.java]: Use `.isEmpty()` instead of comparing size to 0.

File: plantuml-parser-core/src/main/java/com/shuzijun/plantumlparser/core/ClassVoidVisitor.java
Comment (espertus) [inline plantuml-parser-core/src/main/java/com/shuzijun/plantumlparser/core/ClassVoidVisitor.java]: You moved the constructor code before the field code. It should remain after.

File: plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestDefaultConstructor.java
Comment (espertus) [inline plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestDefaultConstructor.java]: It's great that you wrote tests, but you should not duplicate code from `TestRecord`. You should refactor common code. 

I recommend renaming `TestRecord.java`, putting the code currently in `main()` into a new method named something like `testRecordDeclarations()` that gets called from `main()`, and adding a similar method to test default constructors.

File: plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestDefaultConstructor.java
Comment (espertus) [inline plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestDefaultConstructor.java]: Personal preference (not required change): I would make the file extension "uml" or "puml" instead of "txt".

File: plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestDefaultConstructor.java
Comment (espertus) [inline plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestDefaultConstructor.java]: Add to comment that this is when showDefaultConstructors is enabled.

Do something similar for the next comment.

File: plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestDefaultConstructor.java
Comment (espertus) [inline plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestDefaultConstructor.java]: Remove print statement from final version.

File: plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestDefaultConstructor.java
Comment (espertus) [inline plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestDefaultConstructor.java]: Step through in debugger and verify that this shouldn't be "\\n" because the argument to `split()` is a regex.

File: plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestDefaultConstructor.java
Comment (espertus) [inline plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestDefaultConstructor.java]: This is a good method, but you should make some changes...

Helper methods like this should be private.

The first two parameter names are not descriptive.

This is picky (and not a required change), but I would not put the expected output filename as the middle argument. I would put it either first (because it gives the expected value) or last (so it doesn't separate the input file name from the related boolean).

File: plantuml-parser-core/src/main/java/com/shuzijun/plantumlparser/core/ParserConfig.java
Comment (espertus) [inline plantuml-parser-core/src/main/java/com/shuzijun/plantumlparser/core/ParserConfig.java:36]: You need to add a file analogous to [ShowConstructorsOption.java](https://github.com/cathylee12/plantuml-parser/blob/main/plantuml-parser-cli/src/main/java/com/shuzijun/plantumlparser/cli/options/showconstructors/ShowConstructorsOption.java). You should then manually test the new flag (boolean option) from the command line, including that it appears in help messages (presumably `-h`).

**Review by espertus — CHANGES_REQUESTED** [review]
Nice work!

File: plantuml-parser-core/src/main/java/com/shuzijun/plantumlparser/core/ParserConfig.java
Comment (cathylee12) [reply plantuml-parser-core/src/main/java/com/shuzijun/plantumlparser/core/ParserConfig.java:36]: Could you please explain more about this?

File: plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestDefaultConstructor.java
Comment (cathylee12) [reply plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestDefaultConstructor.java]: "\n" works fine to me.

File: plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestDefaultConstructor.java
Comment (cathylee12) [reply plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestDefaultConstructor.java]: I will test with puml and apply to the next pull request.

File: plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestRecordAndDefaultConstructor.java
Comment (espertus) [inline plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestRecordAndDefaultConstructor.java]: I think it would be better to give it a more general name, such as `UmlGenerationTester`.

File: plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestRecordAndDefaultConstructor.java
Comment (espertus) [inline plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestRecordAndDefaultConstructor.java]: Good use of parameter.

File: plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestRecordAndDefaultConstructor.java
Comment (espertus) [inline plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestRecordAndDefaultConstructor.java]: The meaning of the 3 parameters isn't clear. My guess is that uml1 is compared to either uml2 or what is read from umlFilename. If so, those should be two separate methods.

File: plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestRecordAndDefaultConstructor.java
Comment (espertus) [inline plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestRecordAndDefaultConstructor.java]: I don't think there should be different `isUmlEquivalent` methods for records vs. default constructors. There's no reason for them to be different, and that implies that other language features would need their own methods. I think you should probably have:

```
boolean isUmlEquivalent(String uml1, String uml2)
```
and
```
boolean isUmlEquivalent(String uml1, String filename)
```
The second method would call the first one.

File: plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestRecordAndDefaultConstructor.java
Comment (espertus) [inline plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestRecordAndDefaultConstructor.java]: The meaning of `i` is not clear. Maybe call it `numLines`.

File: plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestRecordAndDefaultConstructor.java
Comment (espertus) [inline plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestRecordAndDefaultConstructor.java]: I'd move lines 118-122 into one new method and lines 124-135 into another.

**Review by espertus — CHANGES_REQUESTED** [review]

File: plantuml-parser-cli/build/tmp/compileJava/source-classes-mapping.txt
Comment (espertus) [inline plantuml-parser-cli/build/tmp/compileJava/source-classes-mapping.txt]: This and a few other files need to be removed.

File: plantuml-parser-cli/src/main/java/com/shuzijun/plantumlparser/cli/options/CliOptionsHolder.java
Comment (espertus) [inline plantuml-parser-cli/src/main/java/com/shuzijun/plantumlparser/cli/options/CliOptionsHolder.java:49]: Good. I like how you placed it immediately after ShowConstructorsOption.

File: plantuml-parser-cli/src/main/java/com/shuzijun/plantumlparser/cli/options/showdefaultconstructors/ShowDefaultConstructorsOption.java
Comment (espertus) [inline plantuml-parser-cli/src/main/java/com/shuzijun/plantumlparser/cli/options/showdefaultconstructors/ShowDefaultConstructorsOption.java]: I personally like "sdctr" better than "sdfctr" (since it is more similar to sctr), but it is up to you whether to change it.

File: plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestDefaultConstructor.java
Comment (espertus) [reply plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestDefaultConstructor.java]: Be sure to remove the print statement.

File: plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestDefaultConstructor.java
Comment (espertus) [reply plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/TestDefaultConstructor.java]: This class has code duplicated in `UmlGenerationTester.java`. Let me know if you'd like to discuss the refactoring.

File: plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/UmlGenerationTester.java
Comment (espertus) [inline plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/UmlGenerationTester.java:17]: If you copied code from `TestRecord.java`, you should have `TestRecord` call this code, rather than duplicate it.

**Review by espertus — CHANGES_REQUESTED** [review]

File: plantuml-parser-cli/build/tmp/compileJava/source-classes-mapping.txt
Comment (cathylee12) [reply plantuml-parser-cli/build/tmp/compileJava/source-classes-mapping.txt]: Yes, I added the file path to .gitignore after I pushed. I will delete the files and push again.

File: plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/UmlGenerationTester.java
Comment (cathylee12) [reply plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/UmlGenerationTester.java:17]: My intention was that UmlGenerationTester replaces both TestRecord.java and TestDefaultConstructor.java. My plan was to delete these files, but I left them until I get UmlGenerationTester approved by you. What do you think about this?

File: plantuml-parser-cli/build/tmp/compileJava/source-classes-mapping.txt
Comment (espertus) [reply plantuml-parser-cli/build/tmp/compileJava/source-classes-mapping.txt]: It's always a pain removing gitignore'd files.

File: plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/UmlGenerationTester.java
Comment (espertus) [reply plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/UmlGenerationTester.java:17]: Good to know. Because old files are never truly lost, you can delete files casually in git.

File: plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/UmlGenerationTester.java
Comment (espertus) [inline plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/UmlGenerationTester.java]: Use try-with-resources. https://www.baeldung.com/java-try-with-resources

File: plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/UmlGenerationTester.java
Comment (espertus) [inline plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/UmlGenerationTester.java]: FYI, you can also write `sb.append(cur).append('\n')` (method chaining)

File: plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/UmlGenerationTester.java
Comment (espertus) [inline plantuml-parser-core/src/test/java/com/shuzijun/plantumlparser/core/UmlGenerationTester.java]: You can drop these comments.

FYI, if you want to know my opinion son comments, see https://stackoverflow.blog/2021/12/23/best-practices-for-writing-code-comments/

**Review by espertus — CHANGES_REQUESTED** [review]
One you remove the remaining files from the `build` directory (if you haven't already), you'll be good to go.

**Review by espertus — APPROVED** [review]
Looks great!

**Pull request merged**
